### ISSUE TITLE
ENH: add a comments variable to pandas.DataFrame.to_csv

### ISSUE THREAD

**Issue by ebo** [pandas-dev/pandas#59839]
### Feature Type

- [X] Adding new functionality to pandas

- [ ] Changing existing functionality in pandas

- [ ] Removing existing functionality in pandas

### Problem Description

For many years I have been adding comments to my Pandas generated CSV files by pre--generating the header, writing out the comments, and then appending the pandas to_csv output:

```
df = ....
with open('example.csv', mode='w') as f:
    f.write(','.join(df.keys())+'\n')
    f.write('# First comment\n')
    f.write('# Second comment...\n')

df.to_csv('example.csv', mode='a', index=False)
```
see StackOverflow post for examples: https://stackoverflow.com/questions/29233496/write-comments-in-csv-file-with-pandas

I think it would be easy to add a "comments" variable to to_csv to write out a list of comments in the header prepended with whatever the comment character is set:

```
my_comments = ["First comment",
               "Second comment..."]
df.to_csv('example.csv', comments=my_comments, index=False)
```

and it would produce a CSV similar to:

> Var1, Var2
> \# First comment
> \# Second comment...
> 123, 456

I think this would be an easy add, and I am willing to code it up and make a pull request, but would like to see if there is any buy-in before spending the time.

### Feature Description

It would be nice if pandas to_csv had a comments variable that worked something like:

```
my_comments = ["First comment",
               "Second comment..."]
df.to_csv('example.csv', comments=my_comments, index=False)
```

and it would produce a CSV similar to:

> Var1, Var2
> \# First comment
> \# Second comment...
> 123, 456

### Alternative Solutions

For years I have been adding comments to my Pandas generated CSV files by doing something like:

```
df = ....
with open('example.csv', mode='w') as f:
    f.write(','.join(df.keys())+'\n')
    f.write('# First comment\n')
    f.write('# Second comment...\n')

df.to_csv('example.csv', mode='a', index=False)
```
see StackOverflow post for examples: https://stackoverflow.com/questions/29233496/write-comments-in-csv-file-with-pandas

### Additional Context

People have been using workarounds for a decade or more (see: https://stackoverflow.com/questions/29233496/write-comments-in-csv-file-with-pandas ), and it would be much more elegant if Pandas added the ability to inject comments into a CSV header.

**Comment by thyripian** [comment]
I’d like to take on this issue. After reviewing the codebase, I’ve identified the initial files that would require modification:

- core/generic.py
- io/formats/format.py
- io/formats/csvs.py

Additional changes may be identified during the testing phase. I will also ensure that this feature is backward-compatible and follows pandas’ existing standards for CSV output.

**Comment by ebo** [comment]
So, from your comments are you going to do the coding, or do you want me and send a pull request?  If you are going to do the coding, then please let me know as I would love to be a beta-tester ;-)

**Comment by thyripian** [comment]
Yes, I would do the coding. It would be nice to have a dedicated beta tester!

**Comment by thyripian** [comment]
@ebo 
Ok, I have made the changes and created a new unit test specifically for the new functionality. You can clone the library from my [fork](https://github.com/thyripian/pandas) and test it at your convenience.

**New Arguments:**
- comment : Accepts a list or tuple of comment strings to prepend to the CSV.
- commentchar : A user-defined character for the comment lines (default is "#").

Please let me know if you encounter any issues or have feedback!

**Comment by ebo** [comment]
@thyripian, question: are you one of the pandas development team?

re: testing the fork...
I will be traveling for a funeral tomorrow and will not be able to test this for at least a week.  That said, the additions of the two variables are exactly what I intuited, and I imagine that I will not have any surprises.

More later.

**Comment by thyripian** [comment]
@ebo 

Thank you for your response. No, I’m not part of the pandas development team; this is my first contribution to the project. I appreciate you taking the time to test the changes. Please take all the time you need, especially given the circumstances. I’ll hold off on submitting the PR until you’re able to test and provide feedback. My condolences for your loss, and safe travels.

**Comment by ebo** [comment]
@thyripian, I thought it odd that I made a feature request, and offered to write the code and make a PR, and then you picked it up and implemented it.  Not sure what the official developers will think of that.

Thank you for your sympathies, and well wishing.

**Comment by thyripian** [comment]
@ebo

I’m sorry for any confusion or misunderstanding. If you’d like to implement the feature yourself, please feel free to do so. I only took it on because I was confident I could handle it, but I completely understand if you’d prefer to work on it yourself. Let me know how you’d like to proceed.

Thank you again for your willingness to help and for your feedback, and sorry if I stepped on toes.

**Comment by ebo** [comment]
@thyripian, lets let the devs say how they want things dealt with, and we can sort out the rest.  It is likely something that we can work on together.

**Comment by rhshadrach** [comment]
The standard [RFC 4180](https://www.ietf.org/rfc/rfc4180.txt) does not specify comments. Unless this is widely implemented by CSV readers, and I believe it is not, I don't think pandas should be adding commas.

**Comment by rhshadrach** [comment]
Thanks for the request - it appears to me the current solution is not too onerous for users and there isn't much gained to be worth the maintenance burden of adding this feature. I am currently opposed to adding this to pandas.

**Comment by mroeschke** [comment]
Agreed, that I don't think this is currently worth expanding the API. Thanks for the suggestion but closing

**Comment by ebo** [comment]
Fair enough.  One of the reasons that I suggested it is that pandas has 
implemented a comment character in the reader, but no reciprocal 
functionality to write comments in the writer. I also use this a lot to 
capture model run configurations in ensembles.  Most recently, I was 
working with an ecologist that needed to process some hydrology data 
whose station information looked to be YAML formatted inside the CSV 
comments, which I parsed out.

Since it does not appear to be willingness to support, I'll continue to 
hard code writing the comments instead of wrapping it into 
pandas.to_csv.

Thank you for your time and consideration.
